router/core/websocket.go

```diff
@@ -308,6 +311,44 @@ func (h *WebsocketHandler) handleUpgradeRequest(w http.ResponseWriter, r *http.R
 		return
 	}
 
+	// Authenticate the connection using the initial payload
+	fromInitialPayloadConfig := h.config.Authentication.FromInitialPayload
+	if fromInitialPayloadConfig.Enabled {
+		// Setting the initialPayload in the context to be used by the websocketInitialPayloadAuthenticator
+		r = r.WithContext(authentication.WithWebsocketInitialPayloadContextKey(r.Context(), handler.initialPayload))
+
+		// Later check access control after initial payload is read and set into the context
+		validatedReq, err := h.accessController.Access(w, r)
+		if err != nil {
+			statusCode := http.StatusForbidden
+			if errors.Is(err, ErrUnauthorized) {
+				statusCode = http.StatusUnauthorized
+			}
+			http.Error(handler.w, http.StatusText(statusCode), statusCode)
+			handler.writeErrorMessage(requestID, err)
+			handler.Close()
+			return
+		}
+		handler.r = validatedReq
+
+		// Export the token from the initial payload to the request header
+		if fromInitialPayloadConfig.ExportTokenToRequestHeader.Enabled {
+			var initialPayloadMap map[string]interface{}
+			err := json.Unmarshal(handler.initialPayload, &initialPayloadMap)
+			if err != nil {
+				requestLogger.Error("Error parsing initial payload: %v", zap.Error(err))
+				handler.writeErrorMessage(requestID, err)
+				return
+			}
+			jwtToken, ok := initialPayloadMap[fromInitialPayloadConfig.Key].(string)
+			if !ok {
+				requestLogger.Error("JWT token is not a string")
```

Comment:
should we do `handler.Close()` here?

Reply:
I'd also add writing an error: `handler.writeErrorMessage(requestID, err)`

Reply:
Can you add a test that triggers this code path and expects an error?

Reply:
Actually, looking closer, this code never gets executed, the websocketInitialPayloadAuthenticator already validates the initial payload and the token itself, in its `Authenticate` function. If either of them is not valid it will log the error(s), send an error message and then closes the websocket, [here is the code](https://github.com/osstotalsoft/cosmo/blob/4e9b5784666688e919c1edb219788c96e94675c7/router/core/websocket.go#L322-L330). 

Here is the [payload validation check](https://github.com/osstotalsoft/cosmo/blob/4e9b5784666688e919c1edb219788c96e94675c7/router/pkg/authentication/initial_payload_authenticator.go#L39-L43) and here the [token validation check](https://github.com/osstotalsoft/cosmo/blob/4e9b5784666688e919c1edb219788c96e94675c7/router/pkg/authentication/initial_payload_authenticator.go#L39-L43)

If it's ok with you, I would remove [both "if" statements](https://github.com/osstotalsoft/cosmo/blob/4e9b5784666688e919c1edb219788c96e94675c7/router/core/websocket.go#L336-L349) from `handleUpgradeRequest`.  What do you think?